File: go/discord.go
package common

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
)

type DiscordAdapter struct { Token string; Client *http.Client }

func (a DiscordAdapter) ParseEvent(payload map[string]any) (InboundEvent, error) {
	message := payload; if nested, ok := payload["d"].(map[string]any); ok { message = nested }
	author, _ := message["author"].(map[string]any); user, _ := author["id"].(string); channel, _ := message["channel_id"].(string); if user==""||channel=="" { return InboundEvent{}, fmt.Errorf("Discord message has no channel or author") }
	event:=InboundEvent{Platform:"discord",UserID:user,ReplyTarget:channel}; if text,ok:=message["content"].(string);ok&&text!="" {event.ContentType="text";event.Text=text;return event,nil}; attachments,_:=message["attachments"].([]any);if len(attachments)==0{return InboundEvent{},fmt.Errorf("Discord message has no supported content")}; attachment,_:=attachments[0].(map[string]any);url,_:=attachment["url"].(string);if url==""{return InboundEvent{},fmt.Errorf("Discord attachment has no URL")};mime,_:=attachment["content_type"].(string);event.ContentType=discordContentType(mime,fmt.Sprint(attachment["filename"]));event.MediaURL=url;return event,nil
}
func discordContentType(mime, filename string) string { value:=strings.ToLower(mime+" "+filename);if strings.HasPrefix(value,"image/")||strings.Contains(value,".png")||strings.Contains(value,".jpg"){return "image"};if strings.HasPrefix(value,"audio/"){return "audio"};if strings.HasPrefix(value,"video/"){return "video"};return "file" }
func (a DiscordAdapter) SendReply(event InboundEvent, reply OutboundReply) error { if event.ReplyTarget==""{return fmt.Errorf("Discord reply_target is required")};client:=a.Client;if client==nil{client=http.DefaultClient};for i,message:=range reply.Messages{if i>=10{break};body,_:=json.Marshal(map[string]any{"content":message.Text,"allowed_mentions":map[string]any{"parse":[]string{}}});req,_:=http.NewRequest(http.MethodPost,"https://discord.com/api/v10/channels/"+event.ReplyTarget+"/messages",bytes.NewReader(body));req.Header.Set("Authorization","Bot "+a.Token);req.Header.Set("Content-Type","application/json");resp,err:=client.Do(req);if err!=nil{return err};io.Copy(io.Discard,resp.Body);resp.Body.Close();if resp.StatusCode/100!=2{return fmt.Errorf("Discord API returned %s",resp.Status)}};return nil }

File: php/src/discord.php
<?php

declare(strict_types=1);

/** @return array<string,mixed> */
function discord_parse_event(array $payload): array
{
    $message = $payload['d'] ?? $payload; $userId = $message['author']['id'] ?? ''; $channelId = $message['channel_id'] ?? '';
    if ($userId === '' || $channelId === '') throw new InvalidArgumentException('Discord message has no channel or author');
    $event = ['platform' => 'discord', 'user_id' => (string)$userId, 'reply_target' => (string)$channelId];
    if (!empty($message['content'])) return $event + ['content_type' => 'text', 'text' => $message['content']];
    $attachment = $message['attachments'][0] ?? []; if (empty($attachment['url'])) throw new InvalidArgumentException('Discord message has no supported content');
    $value = strtolower(($attachment['content_type'] ?? '').' '.($attachment['filename'] ?? '')); $type = str_starts_with($value, 'image/') || preg_match('/\.(png|jpe?g|gif|webp)$/', $value) ? 'image' : (str_starts_with($value, 'audio/') ? 'audio' : (str_starts_with($value, 'video/') ? 'video' : 'file'));
    return $event + ['content_type' => $type, 'media_url' => $attachment['url']];
}

function discord_send_reply(array $event, array $reply, string $token): void
{
    $channel = $event['reply_target'] ?? ''; if ($channel === '' || $token === '') throw new InvalidArgumentException('Discord channel and token are required');
    foreach (array_slice($reply['messages'] ?? [], 0, 10) as $message) {
        $body = ['content' => mb_substr((string)($message['text'] ?? ''), 0, 2000), 'allowed_mentions' => ['parse' => []]];
        $ch = adapter_curl_init('https://discord.com/api/v10/channels/'.rawurlencode((string)$channel).'/messages'); curl_setopt_array($ch, [CURLOPT_POST => true, CURLOPT_RETURNTRANSFER => true, CURLOPT_HTTPHEADER => ['Authorization: Bot '.$token, 'Content-Type: application/json'], CURLOPT_POSTFIELDS => json_encode($body, JSON_THROW_ON_ERROR)]); $result = curl_exec($ch); $status = curl_getinfo($ch, CURLINFO_RESPONSE_CODE); if ($result === false || $status < 200 || $status >= 300) throw new RuntimeException('Discord API request failed'); curl_close($ch);
    }
}
